### PR TITLE
Pass arrays from vector type columns unpacked to Python func

### LINKED ISSUE #2577: Query language fails to call python methods properly with grouped columns
The query language is not properly calling python functions when columns are grouped.  The example below illustrates the problem.  Array values seem to be expanded and passed in as separate arguments, instead of being passed in as a single array argument.

```python

from deephaven import time_table

t1 = time_table("00:00:01").update(["TimeBin = upperBin(Timestamp, '00:00:03')", "Person = random() > 0.5 ? `A` : random() > 0.5 ? `B` : `C`"])

t2 = t1 \
    .drop_columns("Timestamp") \
    .select_distinct(["TimeBin", "Person"]) \
    .sort(["TimeBin", "Person"]) \
    .group_by("TimeBin")

# single positional argument
def make_pairs_1(a):
    return 1

# take any inputs as an array
def make_pairs_2(*a):
    return 2

# works
x11 = t1.update("Pair=(int)make_pairs_1(Person)")
# works
x12 = t1.update("Pair=(int)make_pairs_2(Person)")
# fails
x21 = t2.update("Pair=(int)make_pairs_1(Person)")
# works
x22 = t2.update("Pair=(int)make_pairs_2(Person)")
```

### PR DESCRIPTION
**Pull request by jmao-denver**
Fixes #2577

### CODE REVIEW

File: py/server/tests/test_table.py [hunk 1]
```diff
@@ -859,6 +859,33 @@ def test_tree(self):
         tree_table = self.test_table.tail(10).tree(id_col='a', parent_col='c', promote_orphans=True)
         self.assertIsNotNone(tree_table)
 
+    def test_grouped_column_as_arg(self):
```
Comment (chipkent) [inline py/server/tests/test_table.py:862]: I would add another set of test tables using java arrays instead of DB arrays.  e.g. `"Person = new Object[]{...}"` and `"Person = new String[]{...}"` etc.

File: engine/table/src/main/java/io/deephaven/engine/table/impl/lang/QueryLanguageParser.java [hunk 2]
```diff
@@ -944,6 +944,27 @@ public static boolean isTypedVector(Class<?> type) {
     private Expression[] convertParameters(final Executable executable,
             final Class<?>[] argumentTypes, final Class<?>[] expressionTypes,
             final Class<?>[][] parameterizedTypes, Expression[] expressions) {
+
+        // For Python callables, all Vector columns should be converted into arrays
```
Comment (chipkent) [inline engine/table/src/main/java/io/deephaven/engine/table/impl/lang/QueryLanguageParser.java:948]: I have a few questions / comments after reading this:
1) This looks like it converts all vectors to arrays, right?
2) I can see doing the array conversion as being convenient, but I can also see it as being low performance.  Do we understand the tradeoffs of always doing it vs having a method in python that gets called?
3) This looks like nodes are being added to the query call expressions.  Will this change effectively break vectorization for python until we do extensive work to make the rest of vectorization work?
4) Have any tests been run to quantify (3)?

**Review by chipkent — CHANGES_REQUESTED** [review]

File: engine/table/src/main/java/io/deephaven/engine/table/impl/lang/QueryLanguageParser.java (hunk 2 above)
Comment (jmao-denver) [reply engine/table/src/main/java/io/deephaven/engine/table/impl/lang/QueryLanguageParser.java:948]: 1. yes
2. it is actually more performant because JPY knows how to handle Java arrays not vectors. So converting vectors to arrays first should reduce the boundary-crossing when the Python code iterates over the data.
3. Python function calls with array/vector argument can't be vectorized.

File: docker-compose-common.yml [hunk 3]
```diff
@@ -14,26 +14,31 @@ services:
       # with max memory.
       #
       # To turn on debug logging, add: -Dlogback.configurationFile=logback-debug.xml
-      - START_OPTS=-Xmx4g -Ddeephaven.console.type=${DEEPHAVEN_CONSOLE_TYPE} -Ddeephaven.application.dir=${DEEPHAVEN_APPLICATION_DIR}
+      # - JAVA_TOOL_OPTIONS=-Xmx4g -Ddeephaven.console.type=${DEEPHAVEN_CONSOLE_TYPE} -Ddeephaven.application.dir=${DEEPHAVEN_APPLICATION_DIR}
```
Comment (devinrsmith) [inline docker-compose-common.yml:17]: I don't think we want any changes to this file?

File: docker/server-jetty/src/main/docker/Dockerfile [hunk 4]
```diff
@@ -28,13 +28,26 @@ FROM $BASE
 COPY --from=install / /
 VOLUME /data
 VOLUME /cache
-HEALTHCHECK --interval=3s --retries=3 --timeout=11s CMD /opt/grpc_health_probe/grpc_health_probe -addr=localhost:10000 -connect-timeout=10s || exit 1
+HEALTHCHECK --interval=3s --retries=3 --timeout=11s CMD /bin/grpc_health_probe -addr=localhost:10000 -connect-timeout=10s || exit 1
```
Comment (devinrsmith) [inline docker/server-jetty/src/main/docker/Dockerfile:31]: I'm confused why this change was made...

File: docker-compose-common.yml (hunk 3 above)
Comment (chipkent) [reply docker-compose-common.yml:17]: agree

File: docker/server-jetty/src/main/docker/Dockerfile (hunk 4 above)
Comment (chipkent) [reply docker/server-jetty/src/main/docker/Dockerfile:31]: agree

File: engine/table/src/main/java/io/deephaven/engine/table/impl/lang/QueryLanguageParser.java (hunk 2 above)
Comment (chipkent) [reply engine/table/src/main/java/io/deephaven/engine/table/impl/lang/QueryLanguageParser.java:948]: Does the python code see the input as a numpy array or just a buffer that needs to be converted to a numpy array?

File: docker-compose-common.yml (hunk 3 above)
Comment (jmao-denver) [reply docker-compose-common.yml:17]: sorry, dumb mistake

File: engine/table/src/main/java/io/deephaven/engine/table/impl/lang/QueryLanguageParser.java (hunk 2 above)
Comment (jmao-denver) [reply engine/table/src/main/java/io/deephaven/engine/table/impl/lang/QueryLanguageParser.java:948]: The Python code sees it as a JPY wrapped Java array not a numpy array. If the array is a primitive one, its buffer can be used to create a numpy array with zero-copy.

Comment (chipkent) [reply engine/table/src/main/java/io/deephaven/engine/table/impl/lang/QueryLanguageParser.java:948]: This seems very usable without always converting to numpy.

image

My only concern is the lack of vectorization, but I think that is a separate effort.

**Review by chipkent — DISMISSED** [review]

File: engine/table/src/main/java/io/deephaven/engine/table/impl/lang/QueryLanguageParser.java [hunk 5]
```diff
@@ -944,6 +944,27 @@ public static boolean isTypedVector(Class<?> type) {
...
+                                    new NodeList<>(expressions[ei]));
+                    // To prevent JPY from unpacking a non-primitive array to make sure the result array is treated
+                    // as a single argument
+                    if (expressionTypes[ei] == ObjectVector.class) {
+                        expressions[ei] = new CastExpr(
+                                new ClassOrInterfaceType("java.lang.Object"),
+                                expressions[ei]);
+                    }
+                    expressionTypes[ei] = convertVector(expressionTypes[ei],
+                            parameterizedTypes[ei] == null ? null : parameterizedTypes[ei][0]);
```
Comment (rcaudy) [inline engine/table/src/main/java/io/deephaven/engine/table/impl/lang/QueryLanguageParser.java:963]: I'm nervous that in the case of `ObjectVector` we are creating an expression that explicitly conflicts with the expression type. Maybe we need the code to be this:
```suggestion
                    if (expressionTypes[ei] == ObjectVector.class) {
                        expressions[ei] = new CastExpr(
                                new ClassOrInterfaceType("java.lang.Object"),
                                expressions[ei]);
                        expressionTypes[ei] = Object.class;
                    } else {
                        expressionTypes[ei] = convertVector(expressionTypes[ei],
                                parameterizedTypes[ei] == null ? null : parameterizedTypes[ei][0]);
                    }
```
